### ISSUE TITLE
Possible bug in calculation of the value of the function at time t

### ISSUE THREAD

**Issue by mohamad-amin** [google/neural-tangents#126]
Hey! 
I was reading through the code and I noticed that you're using element-wise exponential matrix here:

https://github.com/google/neural-tangents/blob/5f286b7696364217aa4a2d92378aabd0203a791e/neural_tangents/predict.py#L1180

Does this correspond to this equation (9) in the paper?

<img width="1052" alt="Screen Shot 2021-10-05 at 4 55 58 PM" src="https://user-images.githubusercontent.com/5057955/136119563-ab47ea7b-89bd-4de1-8d07-27158893f493.png">

If yes, shouldn't it be [matrix exponential](https://en.wikipedia.org/wiki/Matrix_exponential) instead of element-wise exponential?

Thanks in advance.

**Comment by sschoenholz** [comment]
Great question! We move into a diagonal basis before applying the exponential in which case elementwise exponential and matrix exponential agree. In other words, we write $\exp(At)v = U\exp(Dt)U^Tv$ where $A = UDU^T$ is the diagonalization of $A$.

Let me know if that helps, or if you have any other concerns.

All the best,
Sam

**Comment by mohamad-amin** [comment]
Thanks! Sorry for not paying attention to this.

It makes sense to me now, but I'm a bit concerned about the PSD correction that you apply to the kernel matrix here. In practice, as we care about the converged function, it would be pleasant to take the limit of t in infinite, which translates to choosing a really large t here. However, this correction, combined with multiplication with a really large t might cause approximation issues leading to the final learnt function not being as close as it should be to the finite network.

Sincerely,
Amin

**Comment by romanngg** [comment]
IIUC you'd want to pass `t=None`, which we treat as symbolic infinity and choose a code path that does the simplified expression like Eq 16 in https://arxiv.org/pdf/1902.06720.pdf

**Comment by romanngg** [comment]
As far as I understand, for specific finite width `n`, there will always be a mismatch between your network and its linearization, and we don't have super practical bounds on the magnitude of the mismatch, so I would probably do binary search on width `n` in practice and see if it's large enough to give a close match.

In addition to this, you also need to satisfy conditions of Theorem 2.1 itself to expect that for large `n` the source and the linearized networks will behave closely. Beyond using full-batch gradient descent, checking that your inputs have the proper norm as stated in the theorem, you can also make sure your learning rate is not too big by calling [`nt.predict.max_learning_rate`](https://neural-tangents.readthedocs.io/en/latest/neural_tangents.predict.html#neural_tangents.predict.max_learning_rate). Finally, the theorem also requires that your empirical NTK converges in the limit of large `n`; if you build your network out of `nt.stax` layers, it will be so, but for arbitrary functions it's not guaranteed (although, it should converge for a very wide range of architectures; see https://arxiv.org/pdf/1902.04760.pdf), so unfortunately this is one more way in which it's hard to certify convergence with certainty.

(CC @SiuMath who may know more about convergence and bounding the discrepancy)
